x/tx/signing/aminojson/json_marshal.go

```diff
@@ -17,17 +20,31 @@ type MessageEncoder func(*Encoder, protoreflect.Message, io.Writer) error
 // FieldEncoder is a function that can encode a protobuf protoreflect.Value to JSON.
 type FieldEncoder func(*Encoder, protoreflect.Value, io.Writer) error
 
+// EncoderOptions are options for creating a new Encoder.
+type EncoderOptions struct {
+	FileResolver signing.ProtoFileResolver
+	TypeResolver protoregistry.MessageTypeResolver
+}
+
 // Encoder is a JSON encoder that uses the Amino JSON encoding rules for protobuf messages.
 type Encoder struct {
 	// maps cosmos_proto.scalar -> field encoder
 	scalarEncoders  map[string]FieldEncoder
 	messageEncoders map[string]MessageEncoder
 	fieldEncoders   map[string]FieldEncoder
+	fileResolver    signing.ProtoFileResolver
+	typeResolver    protoregistry.MessageTypeResolver
 }
 
 // NewEncoder returns a new Encoder capable of serializing protobuf messages to JSON using the Amino JSON encoding
 // rules.
-func NewEncoder() Encoder {
+func NewEncoder(options EncoderOptions) Encoder {
```

Comment:
If the options are facultative and do not need to be used at the same time, why not having them variadic?
Now you needed to add an empty struct everywhere.

Reply:
It seems like a misuse of variadic to me since we never want more than one of them.

Reply:
Maybe we could have two constructors:
`NewEncoder`
`NewEncoderWithOptions`

Personally it seems OK to me supply empty options as a default with only one constructor.  Any thoughts @aaronc ?

Reply:
Empty options would almost always be incorrect

Reply:
They're correct in tests.  What are you proposing instead?

Reply:
I propose one `NewEncoder` which takes options.  Empty options are probably only correct in tests with a narrow scope. The AminoJSON encoder needs the HybridResolver from https://github.com/cosmos/gogoproto/pull/62 in production or any and dynamic type marshaling will fail.  The only reason I didn't make HybridResolver the default for `EncoderOptions.FileResolver` is that this would form a dependency on gogoproto, and there currently isn't one.